src/ir/stack-utils.cpp

```diff
@@ -0,0 +1,272 @@
+  StackSignature sig = *this;
+  sig += next;
+  return sig;
+}
+
+StackFlow::StackFlow(Block* block) {
+  // Abstract out the logic for making a pass through the block
+  auto processBlock = [&block](auto process) {
+    for (auto* expr : block->list) {
+      process(expr, StackSignature(expr));
+    }
+    bool unreachable = block->type == Type::unreachable;
+    Type params = unreachable ? Type::none : block->type;
+    process(block, StackSignature(params, Type::none, unreachable));
+  };
+
+  // We need to make an initial pass through the block to figure out how many
+  // values each unreachable instruction produces.
+  std::unordered_map<Expression*, size_t> producedByUnreachable;
+  {
+    size_t stackSize = 0;
+    size_t produced = 0;
+    Expression* lastUnreachable = nullptr;
+    processBlock([&](Expression* expr, StackSignature sig) {
+      // Consume params
+      if (sig.params.size() > stackSize) {
+        // We need more values than are available, so they must come from the
+        // last unreachable.
+        assert(lastUnreachable);
+        produced += sig.params.size() - stackSize;
+        stackSize = 0;
+      } else {
+        stackSize -= sig.params.size();
+      }
+
+      // Handle unreachable or produce results
+      if (sig.unreachable) {
+        if (lastUnreachable) {
+          producedByUnreachable[lastUnreachable] = produced;
+          produced = 0;
+        }
+        assert(produced == 0);
+        lastUnreachable = expr;
+        stackSize = 0;
+      } else {
+        stackSize += sig.results.size();
+      }
+    });
+
+    // Finish record for final unreachable
+    if (lastUnreachable) {
+      producedByUnreachable[lastUnreachable] = produced;
+    }
+  }
+
+  // Take another pass through the block, recording srcs and dests.
+  std::vector<Location> values;
+  Expression* lastUnreachable = nullptr;
+  processBlock([&](Expression* expr, StackSignature sig) {
+    assert((sig.params.size() <= values.size() || lastUnreachable) &&
+           "Block inputs not yet supported");
+
+    // Unreachable instructions consume all available values
+    size_t consumed = sig.unreachable
+                        ? std::max(values.size(), sig.params.size())
+                        : sig.params.size();
+
+    // We previously calculated how many values unreachable instructions produce
+    size_t produced =
+      sig.unreachable ? producedByUnreachable[expr] : sig.results.size();
+
+    srcs[expr] = std::vector<Location>(consumed);
+    dests[expr] = std::vector<Location>(produced);
+
+    // Consume values
+    assert(sig.params.size() <= consumed);
+    size_t unreachableConsumed =
+      consumed > values.size() ? consumed - values.size() : 0;
+    for (Index i = 0; i < consumed; ++i) {
+      if (i < unreachableConsumed) {
+        // This value comes from the polymorphic stack of the last unreachable
+        assert(consumed == sig.params.size());
```

Comment:
If the current instruction is unreachable, it will consume `max(size of stack, size of params)` values. Since we are consuming values from the last unreachable instruction, there must not be enough values available on the stack to satisfy `sig.params`, so we know that `size of stack <= size of params`, so `max(size of stack, size of params) == size of params`. I can expand on this a bit in the code.